### PR TITLE
Fix filter empty Orders

### PR DESCRIPTION
**Pull request by jalbersdorfer**
### 1. Why is this change necessary?
If there are open Orders without Details (Orderlines) - unusual, right - then the Open Orders Overview in Customer Frontend breaks.

### 2. What does this change do, exactly?
Fix filtering out Orders without Details (Orderlines) by delaying the `unset` to after iterating the array.

### 3. Describe each step to reproduce the issue or behaviour.
Produce an Order without any Details (Orderlines) and have a look in the Open Customer Orders Frontend Listing.

### 4. Please link to the relevant issues (if any).
n.a.

### 5. Which documentation changes (if any) need to be made because of this PR?
n.a.

### 6. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have squashed any insignificant commits
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by mitelg** [comment]
hey @jalbersdorfer 
thanks for your PR :slightly_smiling_face: :+1: 

why are there orders without details in the first place? Afaik, this should not be possible in a default installation. Shouldn't it prevented, that orders could be created without details?

I think, your PR fixes only symptoms, but not the cause

**Comment by jalbersdorfer** [comment]
Hi @mitelg
thanks for your quick response.

And es you are right, it should not happen in a default installation.
I observed the behavior in regards to a plugin I'm developing.

The core functionality of the plugin is to combine orders of the same customer within a certain period of time into a single order.
Therefore, I _move_ the Order Details from one to another Order. The original Order Heads get set into a _combined_ state,
so they will not be part any further processing.

The customer on the other Hand has received Order Confirmation E-Mails for the original Orders, and therefore I want them be displayed in the Customer Orders Overview with its correct states and also clearly stated that the original Orders have been merged to another Order.

So that's the root cause is myself. On the other hand, even if it might by default never happen, it is clearly a implementation fault which should be fixed.

I even wanted to first fix this and then talk about changing the behavior to not to remove Orders without details from the List. Which should not be a big thing if it nevertheless should never happen in a default environment.

**Comment by mitelg** [comment]
thanks for the explanation :+1: 

But if you are the reason behind this misbehaviour by yourself, shouldn't you fix it on your site in your plugin? :slightly_smiling_face: 
I'm okay with adding some events or other extensions points, so you can take care of that problem in your plugin properly. what do you think about that?

**Comment by jalbersdorfer** [comment]
In my opinion, I'm not the reason for the misbehavior (bug), I just found it and want to fix it - not to work around it.
The extension Point is already there (`Shopware_Modules_Admin_GetOpenOrderData_FilterResult`), but the array passed to this is broken on receive because of the bug I try to fix with this PR. I'm fully aware that what I do is not a default workflow, and I want to achieve my goal with as less impact to the shopware core as possible. But please understand that this is clearly a bug. Even if it is not me causing the function to run that path, it would break nevertheless. It is a bug. It is fully reproduceable by unit test. It should be fixed. While on the other hand, the implementation details of my plugin are open for any discussion.

**Comment by mitelg** [comment]
hey @jalbersdorfer 

thanks again :slightly_smiling_face: :+1: 

I guess, we can go with your changes. But at least removing of the items should not happen silently. What do you think about some logging here?

I will also make some suggestions regarding your code :wink:

File: engine/Shopware/Core/sAdmin.php [hunk 1]
```diff
@@ -3680,9 +3690,7 @@ private function processOpenOrderDetails(array $orderValue, array $orders, $orde
         );
 
         if (!count($orderDetails)) {
-            unset($orders[$orderKey]);
-
-            return $orders;
+            return false;
```
Comment (mitelg) [inline engine/Shopware/Core/sAdmin.php:3693]: return `null` here, so we can add a proper return type hint to the method

File: engine/Shopware/Core/sAdmin.php [hunk 2]
```diff
@@ -3669,7 +3679,7 @@ private function resendConfirmationMail(array $userInfo, $hash)
      *
      * @param string $orderKey
      *
-     * @return array
+     * @return array|false
      */
     private function processOpenOrderDetails(array $orderValue, array $orders, $orderKey)
```
Comment (mitelg) [inline engine/Shopware/Core/sAdmin.php:3684]:
```suggestion
    private function processOpenOrderDetails(array $orderValue, array $orders, string $orderKey): ?array
```

File: engine/Shopware/Core/sAdmin.php [hunk 3]
```diff
@@ -3669,7 +3679,7 @@ private function resendConfirmationMail(array $userInfo, $hash)
      *
      * @param string $orderKey
      *
-     * @return array
+     * @return array|false
```
Comment (mitelg) [inline engine/Shopware/Core/sAdmin.php:3682]:
```suggestion
```

File: engine/Shopware/Core/sAdmin.php [hunk 4]
```diff
@@ -1353,10 +1354,19 @@ public function sGetOpenOrderData($destinationPage = 1, $perPage = 10)
             $orders[$orderKey]['invoice_shipping'] = $this->moduleManager->Articles()
                 ->sFormatPrice($orderValue['invoice_shipping']);
 
-            $orders = $this->processOpenOrderDetails($orderValue, $orders, $orderKey);
+            $ret = $this->processOpenOrderDetails($orderValue, $orders, $orderKey);
```
Comment (mitelg) [inline engine/Shopware/Core/sAdmin.php:1357]:
```suggestion
            $processedOrders = $this->processOpenOrderDetails($orderValue, $orders, $orderKey);
```

File: engine/Shopware/Core/sAdmin.php [hunk 5]
```diff
@@ -1353,10 +1354,19 @@ public function sGetOpenOrderData($destinationPage = 1, $perPage = 10)
...
+            if ($ret) {
```
Comment (mitelg) [inline engine/Shopware/Core/sAdmin.php:1358]:
```suggestion
            if ($processedOrders !== null) {
```

File: engine/Shopware/Core/sAdmin.php [hunk 6]
```diff
@@ -1353,10 +1354,19 @@ public function sGetOpenOrderData($destinationPage = 1, $perPage = 10)
...
+                $orders = $ret;
```
Comment (mitelg) [inline engine/Shopware/Core/sAdmin.php:1359]:
```suggestion
                $orders = $processedOrders;
```

**Review by mitelg — CHANGES_REQUESTED** [review]
also adding a test would be nice :+1:

**Comment by jalbersdorfer** [comment]
Hi, as you state in your comment

> But at least removing of the items should not happen silently. What do you think about some logging here?

I did not start removing the Empty Order Bodies in the first place. I'm just producing them. The silent removing is the default behavior of the Shopware Core which just has an implementation fault.

I really would like to see the empty Order Heads not getting removed silently, but just kept. 

If you agree, I would change the Implementation to just keep them. - Then most of my changes and of your suggestions would be obsolete.

What do you think?

**Comment by mitelg** [comment]
okay you are right. we can leave the logging out

**Comment by shyim** [comment]
Hey @jalbersdorfer,

this PR seems still incomplete due to requested changes. We would like to let it open for another week and close it if we don't get a response from you.

**Pull request closed without merging**